src/System.Buffers.Experimental/System/Buffers/ReadWriteBytes.cs

```diff
@@ -0,0 +1,322 @@
+                if (advance) position.SetItem(Rest);
+                return (!_first.IsEmpty || _all != null);
+            }
+            if (position.IsEnd)
+            {
+                value = default;
+                return false;
+            }
+
+            var (segment, index) = position.Get<IMemoryList<byte>>();
+
+            if (segment == null)
+            {
+                if (_all == null) // single segment ROB
+                {
+                    value = _first.Slice(index - (int)_totalLengthOrOffset);
+                    if (advance) position = Position.End;
+                    return true;
+                }
+                else
+                {
+                    value = default;
+                    return false;
+                }
+            }
+
+            var memory = segment.Memory;
+
+            // We need to slice off the last segment based on length of this ROB. 
+            // This ROB is a potentially shorted view over a longer segment list.
+            var virtualIndex = segment.VirtualIndex;
+            var lengthOfFirst = virtualIndex - VirtualIndex;
+            var lengthOfEnd = lengthOfFirst + memory.Length;
+            if (lengthOfEnd > Length)
+            {
+                if (advance) position = Position.End;
+                value = memory.Slice(0, (int)(Length - lengthOfFirst));
+                if (index < value.Length)
+                {
+                    if (index > 0) value = value.Slice(index);
+                    return true;
+                }
+                else
+                {
+                    value = Memory<byte>.Empty;
+                    return false;
+                }
+            }
+            else
+            {
+                value = memory.Slice(index);
+                if (advance) position.SetItem(segment.Rest);
+                return true;
+            }
+        }
+
+        public Memory<byte> First => _first;
+
+        internal IMemoryList<byte> Rest => _all?.Rest;
+
+        public long Length => _all == null ? _first.Length : _totalLengthOrOffset;
+
+        private long VirtualIndex => _all == null ? _totalLengthOrOffset : _all.VirtualIndex + (_all.Memory.Length - _first.Length);
+
+        public bool IsEmpty => _first.Length == 0 && _all == null;
+
+        public static ReadWriteBytes Empty => s_empty;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadWriteBytes Slice(Range range)
+        {
+            return Slice(range.Index, range.Length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadWriteBytes Slice(long index, long length)
+        {
+            if (length == 0) return Empty;
+
+            if (_all == null)
+            {
+                if (index > _first.Length) throw new ArgumentOutOfRangeException(nameof(index));
+                if (length > _first.Length - index) throw new ArgumentOutOfRangeException(nameof(length));
```

Comment:
What about index/length being negative?

Reply:
Memory ctor will do the check that.

... this is the never ending dilemma: how much double checking we want to do.

Reply:
Fair point. Why do these checks then? Slice will throw if out of bounds.

Reply:
I added the checks. It helps later in code. At some point we can optimize them out, but I will do it later if needed.